### ISSUE TITLE
Add support for styled-components v4

### ISSUE THREAD

**Issue by kripod** [gatsbyjs/gatsby#7897]
## Summary

I tried to update [a project](
https://github.com/simonyiszk/frt-web/tree/2620007f320bf635dbb0744746d557c119897cb8) to styled-components@next, but exceptions get thrown during `gatsby develop`.

**Comment by DSchau** [comment]
@kripod could you share what those errors are?

Also most helpful is a repo demonstrating the issue, so we can check it out ourselves :)

(Also way to be on the bleeding edge 👍 I _just_ saw the announcement, ha!)

**Comment by kripod** [comment]
@DSchau Unfortunately, I didn’t copy the error message. However, even after disabling `gatsby-plugin-styled-components@next`, the same error was thrown _(by gatsby-ssr if I remember correctly, something like the property “target” cannot be found on a styled component)_.

**Comment by kakadiadarpan** [comment]
Hi @kripod, like @DSchau suggested it would be great if you can provide a demo project. That makes your issue much easier to diagnose.

**Comment by amankkg** [comment]
I have an issue with v4 `createGlobalStyle`, which is a replacement for `injectGlobal`. [Here is my example](https://github.com/amankkg/amank.me/tree/58b5d96ca72de2639cb7fcb49624fc2e593b150f) with latest sc and gatsby. Global style component is [rendered in layout component](https://github.com/amankkg/amank.me/blob/58b5d96ca72de2639cb7fcb49624fc2e593b150f/src/components/layout.js#L27). The warning says next:

> The global style component sc-global-3629742163 was composed and rendered multiple times in your React component tree. Only the last-rendered copy will have its styles remain in <head>.

This warning pops up only on (every) navigation, e.g. on initial render it is OK. But, once this warning appears the global styles become broken.

**Comment by tricoder42** [comment]
I've just tried styled-components v4 with Gatsby v2. In develop everything works fine, but after build the styles are missing completely. No error in console. 

Here's the [repository](https://github.com/tricoder42/tomasehrlich.cz/tree/f2c8718bf95cea89395a78f62cd5a19c2205aa74).

I've downgraded to styled-components v3 and it works.

**Comment by garetmckinley** [comment]
Having the same problem, so I created a repo for it [here](https://github.com/garetmckinley/gatsby-styled-components-v4)

The error I'm seeing is:

```
Unhandled promise rejection TypeError: "React.createContext is not a function"
	js http://localhost:8000/commons.js:48548:21
	js http://localhost:8000/commons.js:47172:30
	__webpack_require__ http://localhost:8000/commons.js:556:12
	fn http://localhost:8000/commons.js:87:20
	js http://localhost:8000/commons.js:2647:26
	__webpack_require__ http://localhost:8000/commons.js:556:12
	fn http://localhost:8000/commons.js:87:20
	js http://localhost:8000/commons.js:1602:15
	__webpack_require__ http://localhost:8000/commons.js:556:12
	fn http://localhost:8000/commons.js:87:20
	js http://localhost:8000/commons.js:2503:36
	__webpack_require__ http://localhost:8000/commons.js:556:12
	fn http://localhost:8000/commons.js:87:20
	js http://localhost:8000/commons.js:2166:22
	__webpack_require__ http://localhost:8000/commons.js:556:12
	fn http://localhost:8000/commons.js:87:20
	js http://localhost:8000/commons.js:946:15
	run http://localhost:8000/commons.js:5674:23
	notify http://localhost:8000/commons.js:5691:31
	flush http://localhost:8000/commons.js:5134:10
```

The exact same repo works fine if you install v3.

**Comment by tricoder42** [comment]
You have Gatsby 1.x which uses older React. You need to use Gatsby 2.x which uses latest React. I don’t think it’s possible to upgrade React in Gatsby without version bump, it’s a breaking change.

**Comment by garetmckinley** [comment]
Woah, that's odd. I swear I had Gatsby v2 installed globally, I'll reinitialize my project and report back.

**Comment by KyleAMathews** [comment]
`gatsby-cli` isn't the same thing as `gatsby`. The CLI just runs whatever version of `gatsby` you have installed in your project.

**Comment by amankkg** [comment]
Thanks to @tricoder42 and @garetmckinley for examples, I've updated and fixed my own example and it seems like there are no errors or malfunctions so far, only that warning in dev mode. Global styles are working even on navigation.

@tricoder42 seems like your example has only one page and if you add another one the warning may occur.
@garetmckinley your example has navigation issue: once navigated to page 2 global styles [seem to be broken](https://prnt.sc/ksbu0p) and dev warning persists.

**Comment by garetmckinley** [comment]
Yup, I can confirm that what @amankkg said regarding global styles is happening, even when deployed.

I created a repo for my new blog [here](https://github.com/garetmckinley/words.garet.io), and I deployed the static site to gh pages [here](http://words.garet.io). I have the prism syntax styles in [my global styles](https://github.com/garetmckinley/words.garet.io/blob/master/src/styles/GlobalStyles.js) and when you navigate to a post, the syntax theme does not appear. As soon as you reload the page on the post, the syntax theme shows up.

I'm going to add in SSR support to my blog and deploy it to my server to see if the styles persist that way. But as it looks now, global styles are broken when routing on static deployments.

**Comment by probablyup** [comment]
I've been looking into this and I think it's because upon switching pages the previous instance doesn't get unmounted until after the next one has been rendered. I think I'll make a change to when this warning is triggered because we just don't want concurrent renders happening.

**Comment by gi-alec** [comment]
Getting an error in every location I import styled-components:
`Module not found: Error: Can't resolve 'styled-components/no-tags'`

Sure enough, checking node_modules/styled-components/no-tags/dist, the folder is empty.

**Comment by probablyup** [comment]
Update your babel plugin please

**Comment by fabe** [comment]
Rendering the `globalStyle` component inside `gatsby-browser.js` and `gatsby-ssr.js` gets rid of the warning when switching pages, but it persists on HMR (fabe/gatsby-universal@26a483b).

**Comment by probablyup** [comment]
Yeah I haven’t cut a .3 formal beta release yet. Probably later today

**Comment by probablyup** [comment]
v4.0.0-beta.3 is out which includes the dev warning fix

**Comment by probablyup** [comment]
General note: `styled-components@next` was an old tag we used for the v2 beta and release. I removed it last night so people won't get confused. Should use `@beta` for v4.

**Comment by garetmckinley** [comment]
Yup, latest versions are working great for me! No issues with global styles disappearing during hmr or routing changes. Nicely done @probablyup

**Comment by probablyup** [comment]
@amankkg update your babel plugin

**Comment by amankkg** [comment]
@probablyup I'm on 1.6.4. Error disappeared only after moving this babel plugin to devDependencies. I just followed this https://github.com/styled-components/styled-components/issues/1969#issuecomment-419991738

If this is expected behavior we have to update the Gatsby's SC tutorial.

**Comment by garetmckinley** [comment]
Ok so I just ran into another issue when dealing with global styles. It seems like there is FOUC that is happening on just the global styles.

I'm using:
- `styled-components@4.0.0-beta.5`
- `gatsby-plugin-styled-components@3.0.0-rc.1`

Here's what the issue looks like: https://imgur.com/a/ublAN8R (ignore the gray flickering in the beginning, video encoding issue)

I throttled the page to make sure it's visible, in real time this only is like a 0.5-1s flicker, but it's very noticeable when on mobile.

The individual component styles render instantly (like the purple tag button you see), but the actual [global styles](https://github.com/garetmckinley/words.garet.io/blob/master/src/styles/GlobalStyles.js) (like page margin) do not appear till after the entire page has rendered without them.

I don't think I'm using it incorrectly, I have my global styles rendered in my [main layout](https://github.com/garetmckinley/words.garet.io/blob/master/src/layouts/Main/index.js#L59) and I also have [`gatsby-plugin-styled-components`](https://github.com/garetmckinley/words.garet.io/blob/master/gatsby-config.js#L12) in my `gatsby-config.js`.

Everything else has been working great on this latest beta

**Comment by probablyup** [comment]
@garetmckinley can you provide a reproduction repo or codesandbox?

**Comment by garetmckinley** [comment]
Yeah, won't be able to till tomorrow afternoon/evening sometime. The blog I linked to is open source if you wanna look at it, but I totally understand if you'd rather wait for a stripped down repro

**Comment by probablyup** [comment]
Yeah it's too difficult to chase things down without a minimized example. It's also a double-check on your end to make 100% sure there isn't a library clash happening, etc.

**Comment by garetmckinley** [comment]
Yeah np, I'll share a link tomorrow afternoon

**Comment by kakadiadarpan** [comment]
Might be related to #8224

**Comment by probablyup** [comment]
The hoisting fix was released in [v4.0.0-beta.8](https://github.com/styled-components/styled-components/releases/tag/v4.0.0-beta.8)

**Comment by kakadiadarpan** [comment]
Thanks for the update @probablyup! @kripod @garetmckinley @amankkg @tricoder42 can you try upgrading to [v4.0.0-beta.8](https://github.com/styled-components/styled-components/releases/tag/v4.0.0-beta.8) and check if it fixes the issue for you?

**Comment by gi-alec** [comment]
Resolved my issue 👍

**Comment by kakadiadarpan** [comment]
We'll be closing this issue as it's resolution is confirmed by @gi-alec and @jameygleason 

Feel free to [open a new one](https://github.com/gatsbyjs/gatsby/issues/new/choose) if you still experience this problem 👍

**Comment by rclai** [comment]
It seems like this is broken now?

Here are my dependencies
```
"dependencies": {
    "babel-plugin-styled-components": "^1.10.0",
    "gatsby": "^2.0.118",
    "gatsby-image": "^2.0.29",
    "gatsby-plugin-manifest": "^2.0.17",
    "gatsby-plugin-offline": "^2.0.23",
    "gatsby-plugin-react-helmet": "^3.0.6",
    "gatsby-plugin-sharp": "^2.0.20",
    "gatsby-plugin-styled-components": "^3.0.5",
    "gatsby-source-filesystem": "^2.0.20",
    "gatsby-transformer-sharp": "^2.1.13",
    "prop-types": "^15.7.1",
    "react": "^16.8.0",
    "react-dom": "^16.8.0",
    "react-helmet": "^5.2.0",
    "styled-components": "^4.1.3"
  }
```

Doing this
```
// global-styles.js
import { createGlobalStyle } from 'styled-components'

export default createGlobalStyle(`
html, body {
  background-color: red;
}
`)

// layout.js
import GlobalStyles from './global-styles'
....
<GlobalStyles />
```

has no effect

**Comment by probablyup** [comment]
You have to put `GlobalStyles` in your React component tree.

**Comment by rclai** [comment]
@probablyup it is.

**Comment by rclai** [comment]
Crap, used `createGlobalStyle` wrong, I wasn't supposed to use parentheses.
